opensearch-operator/pkg/reconcilers/tenant.go

```diff
@@ -0,0 +1,261 @@
+			RequeueAfter: 10 * time.Second,
+		}
+		return
+	}
+
+	// Check cluster ref has not changed
+	if r.instance.Status.ManagedCluster != nil {
+		if *r.instance.Status.ManagedCluster != r.cluster.UID {
+			reason = "cannot change the cluster an tenant refers to"
+			retErr = fmt.Errorf("%s", reason)
+			r.recorder.Event(r.instance, "Warning", opensearchRefMismatch, reason)
+			return
+		}
+	} else {
+		if pointer.BoolDeref(r.updateStatus, true) {
+			retErr = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+					return err
+				}
+				r.instance.Status.ManagedCluster = &r.cluster.UID
+				return r.Status().Update(r.ctx, r.instance)
+			})
+			if retErr != nil {
+				reason = fmt.Sprintf("failed to update status: %s", retErr)
+				r.recorder.Event(r.instance, "Warning", statusError, reason)
+				return
+			}
+		}
+	}
+
+	// Check cluster is ready
+	if r.cluster.Status.Phase != opsterv1.PhaseRunning {
+		r.logger.Info("opensearch cluster is not running, requeueing")
+		reason = "waiting for opensearch cluster status to be running"
+		r.recorder.Event(r.instance, "Normal", opensearchPending, reason)
+		retResult = ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 10 * time.Second,
+		}
+		return
+	}
+
+	r.osClient, retErr = util.CreateClientForCluster(r.ctx, r.Client, r.cluster, r.osClientTransport)
+	if retErr != nil {
+		reason = "error creating opensearch client"
+		r.recorder.Event(r.instance, "Warning", opensearchError, reason)
+		return
+	}
+
+	// Check tenant state to make sure we don't touch preexisting tenants
+	if r.instance.Status.ExistingTenant == nil {
+		var exists bool
+		exists, retErr = services.TenantExists(r.ctx, r.osClient, r.instance.Name)
+		if retErr != nil {
+			reason = "failed to get tenant status from Opensearch API"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		if pointer.BoolDeref(r.updateStatus, true) {
+			retErr = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+					return err
+				}
+				r.instance.Status.ExistingTenant = &exists
+				return r.Status().Update(r.ctx, r.instance)
+			})
+			if retErr != nil {
+				reason = fmt.Sprintf("failed to update status: %s", retErr)
+				r.recorder.Event(r.instance, "Warning", statusError, reason)
+				return
+			}
+		} else {
+			// Emit an event for unit testing assertion
+			r.recorder.Event(r.instance, "Normal", "UnitTest", fmt.Sprintf("exists is %t", exists))
+			return
+		}
+	}
+
+	// If tenant is existing do nothing
+	if *r.instance.Status.ExistingTenant {
+		reason = opensearchTenantExists
+		return
+	}
+
+	tenant := requests.Tenant{
+		Description: r.instance.Spec.Description,
+	}
+
+	shouldUpdate, retErr := services.ShouldUpdateTenant(r.ctx, r.osClient, r.instance.Name, tenant)
```

Comment:
are you reaching that step? you dont exit on the return on the .ExistingTenant part .

Reply:
Yes we will reach the step, if the tenant is created by the operator